### ISSUE TITLE
nested routes

### ISSUE THREAD

**Issue by r3wt** [ReactTraining/react-router#5814]
there should be nested routes, you fucking idiots.

**Comment by josepot** [comment]
Like [this](https://reacttraining.com/react-router/web/example/recursive-paths)? Nested routes are 100% supported by `react-router`.

**Comment by r3wt** [comment]
@josepot very funny. i meant nesting `Routes` within each other. for example i wanted to have `Route` render an `AppComponent` but only for a certain set of routes. Then, nested routes for the pages in that `AppComponent`. now this is no longer possible. it just seems to not make any sense to me. now i would have to hide my routes within the AppComponent, thus hiding the route logic from the higher order component.

**Comment by r3wt** [comment]
ie something like
```js
<Route path='/blog' component={Blog}>
   <IndexRoute component={BlogList} />
   <Route path='/archive/:month' component={BlogList} />
   <Route path='/:slug' component={BlogPost} />
</Route>
```

the stupidity of the package maintainer is mind boggling

**Comment by josepot** [comment]
Again: did you even bother having a look at the docs?

```js
<Route path='/blog' render={({match}) =>
  <BlogLayout>
     <IndexRoute component={BlogList} />
     <Route exact path={`${match.url}/archive/:month`} component={BlogList} />
     <Route exact path={`${match.url}/:slug`} component={BlogPost} />
  </BlogLayout>
}/>
```

If you don't like the API of this library you have a few options:
- Fork it, change it and use your fork
- Make your own library
- Suggest what you think is a better API and try to make constructive contributions (I would disagree with you on this particular topic)

I don't think that insulting and trashing is an acceptable option.

**Comment by r3wt** [comment]
@josepot that looks horrendous, but i guess its passable. why the change, is it to avoid passing the child component as children in props? the course i took said that it was like i give in my example, then i am wasting time trying to figure out why it wasn't working. that's why i am angry. it looks like a dumb change and there is no reason given

**Comment by r3wt** [comment]
@josepot further, wasn't it obvious what i thought the better API was from my original post? the change in API seems superfluous to me, but perhaps you wouldn't care to explain why?

**Comment by josepot** [comment]
> will your example work if the nested routes are inside of a Switch component

I've updated my example. Now it uses a `Switch` component and yep, it should work.

> it looks like a dumb change and there is no reason given

> the change in API seems superfluous to me, but perhaps you wouldn't care to explain why

Surprisingly, the answer to these questions can be found [here](https://reacttraining.com/react-router/core/guides/philosophy)... In the docs. 😄 

I mean, man it seems to me that you learned to use this library by taking an outdated course and that's on you... libraries improve their API over time, react-router uses [semver](https://semver.org/) so it's totally fair to make major changes on the API from major version to major version.

> wasn't it obvious what i thought the better API was from my original post?

It wasn't to me. It was not until v4 of this library that I started using `react-router`... I didn't like the API of the previous versions, and since I use redux I just consumed [`history`](https://www.npmjs.com/package/history) directly through `redux-saga`. I'm serious, I'm zero familiar with the previous versions of this library... Every time I looked at the API of the previous versions I was like: "nah!, not for me!" Until v4, then I was like: "yep, now we are talking!".

**Comment by r3wt** [comment]
hmm, after reading that i guess it makes some semblance of sense, but would have been nice if they would have left the previous versions as is since it suits the most use cases.

**Comment by josepot** [comment]
> would have been nice if they would have left the previous versions as is

:joy: Man, you are free to use v3 if you want:

```
npm install react-router@v3
```
 
the previous API will work just fine with the old version...

**Comment by r3wt** [comment]
@josepot Honest question, how many people do you think look at the following code snippet as a desirable way to write code? Wouldn't this same example be arguably more readable(and drastically less code to boot) with RRv3?

```js
import React from 'react'
import {
  BrowserRouter as Router,
  Route,
  Link
} from 'react-router-dom'

const BasicExample = () => (
  <Router>
    <div>
      <ul>
        <li><Link to="/">Home</Link></li>
        <li><Link to="/about">About</Link></li>
        <li><Link to="/topics">Topics</Link></li>
      </ul>

      <hr/>

      <Route exact path="/" component={Home}/>
      <Route path="/about" component={About}/>
      <Route path="/topics" component={Topics}/>
    </div>
  </Router>
)

const Home = () => (
  <div>
    <h2>Home</h2>
  </div>
)

const About = () => (
  <div>
    <h2>About</h2>
  </div>
)

const Topics = ({ match }) => (
  <div>
    <h2>Topics</h2>
    <ul>
      <li>
        <Link to={`${match.url}/rendering`}>
          Rendering with React
        </Link>
      </li>
      <li>
        <Link to={`${match.url}/components`}>
          Components
        </Link>
      </li>
      <li>
        <Link to={`${match.url}/props-v-state`}>
          Props v. State
        </Link>
      </li>
    </ul>

    <Route path={`${match.url}/:topicId`} component={Topic}/>
    <Route exact path={match.url} render={() => (
      <h3>Please select a topic.</h3>
    )}/>
  </div>
)

const Topic = ({ match }) => (
  <div>
    <h3>{match.params.topicId}</h3>
  </div>
)

export default BasicExample
```

What if i liked and wanted Dynamic Routing, just with Nested Routes?

**Comment by timdorr** [comment]
@r3wt I think what you're looking for here is [react-router-config](https://github.com/ReactTraining/react-router/tree/master/packages/react-router-config). It provides the config-based structure you may have been used to in previous versions of the library. There are some differences, but it's essentially the same. 

The biggest change we made with v4 was treating `<Route>` as an actual component. Before, it was simply a configuration. [It didn't even render!](https://github.com/ReactTraining/react-router/blob/v3/modules/Route.js#L34-L39) Instead, we read `children` and looked directly at their props. This also meant we had to emulate and duplicate a lot of the React lifecycle, but with tons of caveats and non-standard behavior.

Under the new architecture, we're able to maintain the same React semantics you're already used to, and take on new ones as they are added (e.g., async rendering and the in-progress context API changes).  The code base is now dramatically smaller, simpler, and easier to work with. There are certainly some holes and changes that put work onto your plate, but that work should hopefully be more approachable and give you more power in how you want to build your application.

BTW, I would encourage you to exercise some more tact in the future. I don't care about profanity (it's fucking great!), but the sentiment that we're cowboy coding our way through building this library is patently false. We want to incorporate as many ideas and opinions as we can, but they are generally easier to adopt when they aren't as rough around the edges as yours. Hopefully, you can take that as a friendly tip in the future 👍

**Comment by r3wt** [comment]
> @r3wt I think what you're looking for here is react-router-config. It provides the config-based structure you may have been used to in previous versions of the library. There are some differences, but it's essentially the same.

Not now, but it will be useful in the future. I actually never used react before a week or so ago, this is all very new to me, and when it comes to this router, things that should be easy are hard and vice versa. sensible defaults are not to be found. needing `exact` for an exact match on the regexp, wtf is that about? craziness imo. I mean at this point, maybe you can direct me to your crack dealer(s), since you seem to be smoking the good stuff.

> The biggest change we made with v4 was treating <Route> as an actual component. Before, it was simply a configuration. It didn't even render! Instead, we read children and looked directly at their props. This also meant we had to emulate and duplicate a lot of the React lifecycle, but with tons of caveats and non-standard behavior.

I still wish you could have left the nesting behavior for those who desire it. I mean i could downgrade to RRv3, but then i lose out on cool features of RRv4. I actually am not so opposed to the suggestion of @josepot in his example with the render syntax, though i think it is ugly to look at and don't like having to interpolate match.url but whatever. one thing i am confused at is do, nested `Switch` cascade?(not sure if right word but bare with me)

```js
<BrowserRouter>
    <div>
        <Route component={Dialog} />
        <AppLayout>
            <Switch>
                <Route exact path='/' component={Index} />
                <Route exact path='/about' component={About} />
                <Route path='/blog' render={({match}) =>
                    <BlogLayout>
                        <Switch>
                            <Route exact path={`${match.url}/search`} component={BlogList} />
                            <Route exact path={`${match.url}/archive/:month`} component={BlogList} />
                            <Route exact path={`${match.url}/topics/:topic`} component={BlogList} />
                            <Route exact path={`${match.url}/:slug`} component={BlogPost} />
                        </Switch>
                    </BlogLayout>
                />
                
                <Route path='/account' render={({match}) =>
                    <AccountLayout>
                        <Switch>
                            <Route exact path={`${match.url}/content`} component={AccountContentList} />
                            <Route exact path={`${match.url}/profile`} component={AccountProfile} />
                        </Switch>
                    </AccountLayout>
                />
                <Route component={NotFound} />
            </Switch>
        </AppLayout>
    </div>
</BrowserRouter>
```
Would this work as expected. If no matches, even in the nested Routes render'd `Switch`, it will eventually fallover to the NotFound component?


> BTW, I would encourage you to exercise some more tact in the future. I don't care about profanity (it's fucking great!), but the sentiment that we're cowboy coding our way through building this library is patently false. We want to incorporate as many ideas and opinions as we can, but they are generally easier to adopt when they aren't as rough around the edges as yours. Hopefully, you can take that as a friendly tip in the future 👍

You're absolutely right. I was irritated that i had to rewrite my code, so i came in here looking to project that anger on the source of my woe.

**Comment by timdorr** [comment]
I think react-router-config is going to be where you're looking for. You can certainly make that work with the component structure, but resolving to a single `<Route>` will be easier with rrc. It may be a little rough around the edges, but it should be able to get you where you need to go. And anywhere you encounter that roughness, know that we *love* PRs 😄

**Comment by thejameskyle** [comment]
http://thejameskyle.com/dear-javascript.html

**Comment by r3wt** [comment]
> I think react-router-config is going to be where you're looking for. You can certainly make that work with the component structure, but resolving to a single <Route> will be easier with rrc. It may be a little rough around the edges, but it should be able to get you where you need to go. And anywhere you encounter that roughness, know that we love PRs 😄

😞 😢

**Comment by r3wt** [comment]
I will try to suck it up and make the best of what is available in the current implementation of the router

**Comment by iamtommcc** [comment]
@r3wt I am surprised no one has outright said it yet, but your attitude stinks and I am amazed that people had the patience to actually lend you a hand here.

I understand your frustration, but maybe in the future have some empathy for others and try to not be so rude?

Open source projects owe you nothing. They are maintained by volunteers - a.k.a "humans".

**Comment by r3wt** [comment]
@josepot fuck you you little canuck. go to hell

**Comment by r3wt** [comment]
Oh man, what a fool i made of myself. I am deeply sorry to all of you. Happy friday and thanks for the great project (still disagree but nonetheless, i am sorry for my behavior)

**Comment by catamphetamine** [comment]
cyka blyat idi nahui﻿

**Comment by r3wt** [comment]
ok, so about 20 days later, i completely have changed my tune. Awesome library.

**Comment by r3wt** [comment]
@wdjungst that seems like a pretty pretentious thing to say, but whatever. have a nice night.

**Comment by catamphetamine** [comment]
@wdjungst Yeah, Einstein

**Comment by catamphetamine** [comment]
@wdjungst Oh, and you're that man that took `react-project` npm package name.
Fuck you then.

**Comment by wdjungst** [comment]
Took? No the project was handed down to me and 2 weeks later create-react-app was published. Again this is not how you open source. Proven by the edited original post and the replies from you.

**Comment by r3wt** [comment]
@timdorr I'm sorry to bother you man, but you might want to lock this thread again. It seems people can't resist coming in and giving their opinions, trying to pick a fight. Ultimately, I am the one who started it and Am responsible for that. So once again, please accept my sincere apology for the time you have had to waste due to the ignorant comments i made in this thread.

**Comment by catamphetamine** [comment]
Nooooooooooo

![](https://i.pinimg.com/originals/ee/76/2e/ee762ec1372979c708658b5cb3fea175.jpg)

**Comment by timdorr** [comment]
Well, this is going swimmingly. Can I at least get my morning coffee in first?
